### PR TITLE
fix: Context cancellation not handled

### PR DESCRIPTION
**Pull request by fortyanov**

### CODE REVIEW

File: retry.go [hunk 1]
```diff
@@ -41,7 +42,7 @@ const DefaultRetryCap = time.Second * 30
 
 // newRetryTimer creates a timer with exponentially increasing
 // delays until the maximum retry attempts are reached.
-func (c Client) newRetryTimer(maxRetry int, unit time.Duration, cap time.Duration, jitter float64, doneCh chan struct{}) <-chan int {
+func (c Client) newRetryTimer(ctx context.Context, maxRetry int, unit time.Duration, cap time.Duration, jitter float64, doneCh chan struct{}) <-chan int {
```
Comment (vadmeste) [inline retry.go]: It is also possible to remove doneCh here and use context instead

**Comment by fortyanov** [comment]
@vadmeste hello. I made changes what did you say about.

File: retry.go [hunk 2]
```diff
@@ -69,14 +70,13 @@ func (c Client) newRetryTimer(maxRetry int, unit time.Duration, cap time.Duratio
 	go func() {
 		defer close(attemptCh)
 		for i := 0; i < maxRetry; i++ {
+			attemptCh <- i + 1
+
```
Comment (harshavardhana) [inline retry.go]: attemptCh should be kept as part of select otherwise your doneCh is never triggered..

Comment (fortyanov) [reply retry.go]: I remove  doneCh because it not needer any more. Context is uses for breaking retries now.

Comment (harshavardhana) [reply retry.go]: No the correct code should be to keep the attemptCh as part of the select @fortyanov - it won't work as expected for ctx.Done() is called.

Comment (harshavardhana) [reply retry.go]: It should be 

```
select {
   case attemptCh <- i+1:
   case <-time.After(exponentialBackoffWait(i)):
   case ctx.Done():
            return
}
```

Comment (fortyanov) [reply retry.go]: I made this changes.

File: retry.go [hunk 3]
```diff
@@ -70,13 +71,11 @@ func (c Client) newRetryTimer(maxRetry int, unit time.Duration, cap time.Duratio
 		defer close(attemptCh)
 		for i := 0; i < maxRetry; i++ {
 			select {
-			// Attempts start from 1.
 			case attemptCh <- i + 1:
-			case <-doneCh:
-				// Stop the routine.
+			case <-time.After(exponentialBackoffWait(i)):
```
Comment (vadmeste) [inline retry.go]: We can send wrong attempt number here after some timeout.

The better way is:
```
select {
case attemptCh <- i+1:
case <-ctx.Done():
   return 
}

case {
case <-time.After(exponentialBackoffWait(i)):
case <-ctx.Done():
   return
}
```

Comment (harshavardhana) [reply retry.go]: This is the ideal fix

**Comment by fortyanov** [comment]
@vadmeste @harshavardhana @nitisht made new PR because recreate fork for fresh minio version.

**Pull request closed without merging**